### ISSUE TITLE
binary_crossentropy possible numerical instability

### ISSUE THREAD

**Issue by ikvision** [keras-team/keras#9247]
As detailed in Deep Learning textbook by Ian Goodfellow, Yoshua Bengio and Aaron Courville, chapter 4, calculating the loss from logits is more numerically stable compare to calculating it from the probability output.
It seems the default binary_crossentropy in Keras uses probability as input:  from_logits=False
`
   def binary_crossentropy(output, target, from_logits=False):

    if not from_logits:

    # transform back to logits

    epsilon = _to_tensor(_EPSILON, output.dtype.base_dtype)

    output = clip_ops.clip_by_value(output, epsilon, 1 - epsilon)

    output = math_ops.log(output / (1 - output))

   return nn.sigmoid_cross_entropy_with_logits(labels=target, logits=output)
`

The clipping operation avoids overflow, but I think it can reduces the numerical precision.
